### PR TITLE
fix(filter,pipeline): type-coerce scanned finding fields, count exclusion eliminations, disclose the discarded resume (#266, #267, #268)

### LINKED ISSUE #266: apply_exclusions twin divergence on non-string title/description (Python renders "None", JS renders empty)
## Problem

`scripts/filter_findings.py` builds the exclusion scan string with
`f"{finding.get('title','')}..."`, so `title: None` contributes the literal string `None`;
`workflows/src/filterFindings.js` uses `finding.title || ''`, contributing nothing. A pattern
`None` therefore eliminates in the Python twin and keeps in the JS twin (executed 2026-08-31,
PR #264 round-1 red-team). Only `suggestion` is explicitly type-coerced and fixture-pinned
(`exclusions/suggestion_match` carries the four-spelling pattern list for exactly this
class); no parity fixture carries a non-string title or description, so the divergence is
live and unpinned.

## Fix shape

Decide one coercion rule (the string-typed rule the citation fields and `suggestion` already
use is the natural pick — but it is a Python-side behavior change), apply it to both twins,
and add a parity fixture with `title: null` / `description: null` and the four-spelling
pattern list. Python twin has no live consumer, so this is parity hygiene, not a live recall
change.

Surfaced by the PR #264 round-1 red-team.

### LINKED ISSUE #267: exclusion eliminations are invisible in the stats surface
## Problem

Every sibling eliminator in the filter stats block has a count key (`injections_removed`,
`suggested_fix_codes_removed`, `title_matches_eliminated`, ...) but `apply_exclusions`
contributes none: an exclusion-eliminated finding is visible only inside the persisted
`eliminated` array. A user ignore pattern that eliminates a large slice of a run leaves no
trace in the compact return the operator and Phase 8 read.

## Fix shape

Add an `exclusions_removed` count to both twins' stats (one edit each, mirrors the existing
family) and thread it through the receipt/parity comparator like its siblings. Suites-only.

Surfaced by the PR #264 round-1 red-team.

### LINKED ISSUE #268: top-level garbage checkpoints silently discard the resume and re-run the whole review with no disclosure
## Problem

`readCheckpoints` resolves a garbage top-level `checkpoints` value (a string, an array —
arrays since PR #264 — or a wrapper whose `.phases` is not a plain object) to `{}`: the run
dispatches everything fresh, `ok:true`, correct answer — and the operator's resume intent is
silently discarded, with a full re-billed run and no gap saying so. The inertness itself is
the decided F4-3-adjacent behavior (a stamped `checkpoints: null` must stay disclosure-free);
the open question is whether a NON-null garbage value deserves a disclosure gap, since #248's
premise is that a malformed checkpoint must not fail invisibly, and PR #264's shape gate only
covers malformed values inside a well-formed map.

## Fix shape

Decide: either ratify the silent re-run (cheap, document it at `readCheckpoints`) or push a
`checkpoint-shape:`-family disclosure gap when a non-null `checkpoints` resolves to the empty
map. If disclosed, keep `checkpoints: null` exempt (the F4-3 decision). Suites-only.

Surfaced by the PR #264 round-2 red-team.

### PR DESCRIPTION
**Pull request by leehopper**
Closes #266. Closes #267. Closes #268. (#265 is deliberately NOT closed — it is re-scoped by comment; see below.)

## What this does

**#266 — typed-field coercion across the filter twins, fixed class-wide.** The issue named one site (`apply_exclusions` title/description: Python's f-string renders a stamped `None` as the literal string "None" while JS's `|| ''` renders nothing). Recon and two review rounds found the class was much wider, so the rule shipped is: *a scanned finding field contributes a value of its expected type or that type's default — never a stringified null, never a crash*. Shared helpers (`_as_text`/`_as_confidence`, `asText`/`asConfidence`) applied at every site, both twins:

- `apply_exclusions` title/description
- the injection scan core: title/description/file (Python previously **crashed** on a null title at heuristic 7; `file: null` produced a live `elimination_reason` divergence), confidence (Python crashed at heuristic 4; a numeric-string `"90"` bypassed it in JS)
- `detect_disagreement` suppression rules 1 AND 2 (Python crashed; JS silently suppressed on an array title — a live recall divergence)
- `_route_by_dimension` and its duplicate `_is_test_correctness_finding` (live report-destination divergence on a non-string title)
- `consolidate_cross_agent`'s winner key (`description: null` crashed **both** twins through the live `applyFilterPipeline` path)
- `apply_threshold_filter` confidence + severity (null confidence crashed the Python CLI before anything else ran; null severity crashed both twins)

`_as_confidence` excludes bools and maps NaN to 0 in both twins (guards mutation-pinned). Seven new parity fixture groups pin the crash oracles and the cross-language discriminators (array-typed companions where a bare null is inert in one twin); every pre-existing golden is byte-identical.

**#267 — `exclusions_removed` stat, both twins.** Placed adjacent to `injections_removed` in both stats literals and added to the Python module's Output-JSON-schema docstring (its only written contract). Verified against every consumer: the stats object flows opaquely to `reportInput.stats.filter` and the envelope; nothing enumerates keys. Tests drive the real entry points (`main()` with a temp REVIEW.md ignore pattern; `applyFilterPipeline` directly).

**#268 — a discarded resume now discloses.** `readCheckpoints`' unwrap is hoisted to `unwrapCheckpointMap` (signature and return unchanged), and `runWith` pushes one `checkpoint-discarded:` gap when `args.checkpoints` was supplied, non-null, and unusable (`'garbage'`, `[]`, `{phases: null|'x'|[]}` — shapes with no legitimate producer). The gap names the failing position (`checkpoints` vs `checkpoints.phases`) and shape, rides both exits plus the checkpoint-shape reject envelope, and fires exactly once. `checkpoints: null` and absence stay silent — the F4-3 decision is untouched (its test is byte-identical to main). Degrade-and-disclose, not a refusal: PR #264's oracles pin that garbage top-levels dispatch the full fresh pipeline `ok:true`, and a validateArgs refusal would reverse that just-ratified behavior while still missing wrapper shapes. Registered in `docs/machine-parsed-strings.md` (no G3 `_DEGRADE_RE` substrings).

**#265 — deferred, no code.** The pre-implementation red-team refuted the issue's causal premise: every boilerplate `claude_md_rule` in the measurement corpus predates the #218 conditional schema (newest corpus run 2026-08-18; the requirement shipped 2026-08-25), so "schema-required forces it" is unestablished — and the proposed rules-presence signal misses REVIEW.md and code comments, the sources live convention findings actually cite (it would have dropped the schema floor on legitimately-cited findings, with fail-open collector paths on top). The corrected record and re-scope (post-#218 measurement first; any fix targets the prose/emission force) is posted on #265.

## Validation

- Process: 3-reader recon → design memo → 2-opus red-team (REJECT + REVISE; killed the #265 implementation and widened #266) → 2 parallel sonnet implementers + validator (validator caught and fixed one unpinned JS mutation site) → 2-opus adversarial review (both FIX_FIRST; four remaining coercion sites found by execution) → fix round → all gates re-green.
- Suites: 1147 JS / 1828 + 840 Python / 553 bench, green. Coverage: JS 99.08 / 87.99 / 98.55 (floors 98.1 / 86.6 / 97.6), scripts 93.98 (floor 92.9), bench 88.48 (floor 87.5). Bundle rebuilt, second build byte-identical; pre-commit clean.
- Mutation ledgers across all rounds: ~35 mutations, every one singly applied, watched red at a named test/fixture, byte-verified restore, `__pycache__` purged around Python mutations. Notable: the validator proved one fixture non-discriminating for two JS sites and extended it before trusting it.
- Floors: JS branches headroom measures 1.30pp locally (main: 87.51 → 87.99) — a ratchet is owed from this PR's own CI measurement per the AGENTS.md policy and lands as a follow-up commit on this branch once CI reports.

Suites-only per the wave rule. Recall-relevant deltas for the wave-end paired mini: none intended — the coercion changes behavior only for non-string/non-numeric field values the finding schema forbids upstream, plus the JS `"90"`-string threshold bypass now correctly eliminates (previously passed as a string comparison).

Adjacent issue filed from review: a fed-back `{completed, truncated: true}` return silently re-runs everything (the one operator-plausible garbage shape the #268 predicate cannot see — it unwraps to an inert bare map).

🤖 Generated with [Claude Code](https://claude.com/claude-code)

https://claude.ai/code/session_01JnbCPJi3FWXb1JdrRy17HE

---

> [!NOTE]
> **Medium Risk**
> Touches core filter and pipeline resume paths where malformed checkpoint or finding JSON could change elimination, routing, or gap output; mitigated by extensive parity and mutation tests but behavior shifts for edge-case field types.
> 
> **Overview**
> Hardens the **Python and JS filter twins** so scanned finding fields (`title`, `description`, `file`, `severity`, `confidence`) go through shared coercion (`_as_text` / `_as_confidence` and `asText` / `asConfidence`) instead of raw `.get` / template reads. That stops **TypeError crashes** on explicit `null`, stops **literal "None"/"null"** from polluting exclusion/injection scans, and aligns behavior for **non-string values** (e.g. numeric-string `"90"` no longer bypasses thresholds; array titles no longer leak into keyword routing). The same helpers are wired through threshold, injection, exclusions, disagreement suppression, routing, consolidation, and related paths in `scripts/filter_findings.py` and `workflows/src/filterFindings.js` (bundled to `pipeline.js`).
> 
> Adds **`stats.exclusions_removed`** on both filter pipelines (#267). When **`args.checkpoints` is present but unusable**, `runWith` now pushes a **`checkpoint-discarded:`** gap (degrade-and-disclose, run still proceeds fresh) via shared **`unwrapCheckpointMap`** and **`checkpointDiscardGap`** in `workflows/src/stages.js`, including on checkpoint-shape reject envelopes; the string is registered in **`docs/machine-parsed-strings.md`**.
> 
> **CI/docs housekeeping:** JS branch coverage floor **86.6 → 87.0** in `.github/workflows/ci.yml` and `AGENTS.md`, plus a small **`AGENTS_SET_BUDGET_BYTES`** bump. Large parity fixtures and unit tests cover the coercion oracles and discard shapes.
> 
> <sup>Reviewed by [Cursor Bugbot](https://cursor.com/bugbot) for commit 25089a2281f650a043a1c5f49623924c7f46117f. Bugbot is set up for automated code reviews on this repo. Configure [here](https://www.cursor.com/dashboard/bugbot).</sup>

### CODE REVIEW

File: AGENTS.md [hunk 1]
```diff
@@ -76,12 +76,13 @@ LCOV="$(mktemp -d)/js-coverage.lcov" && node --test --experimental-test-coverage
...
+88.29); JS 98.1 / 86.9 / 97.6 (lines pinned
 2026-08-03 from first green CI: 98.61; branches/functions raised 2026-08-18
 from the #62 PR measurement: 86.7/98.4, then 2026-08-26 to 98.1/86.3/97.5 from the #249 PR
 CI measurement: 99.01/87.25/98.48, then branches/functions 2026-08-27 to 86.5/97.6 from
 the #251 PR CI measurement: 99.02/87.47/98.51, then branches 2026-08-27 to 86.6 from
-the #255 PR CI measurement: 87.60). Policy: a floor sits no more than 1.0 pp below the CI
+the #255 PR CI measurement: 87.60, then branches 2026-08-31 to 86.9 from the #269 PR CI
+measurement: 87.99). Policy: a floor sits no more than 1.0 pp below the CI
 measurement for that gate; lower a floor only in the PR that causes the drop,
 reason in the body; raise when measured headroom exceeds 1.0 pp. A sudden
```
Comment (leehopper) [inline AGENTS.md]: **💡 [LOW] JS branches coverage floor (86.9) sits 1.09pp below its own cited CI measurement (87.99), violating the stated ≤1.0pp policy**

AGENTS.md's own floor-ratchet policy requires a floor to sit no more than 1.0 pp below the CI measurement that justifies it. This PR raises the JS branches floor to 86.9 citing the #269 PR CI measurement of 87.99 -- a 1.09pp gap. Every prior entry in this same history paragraph (86.6 from 87.60, 86.5 from 87.47, etc.) respects the ≤1.0pp rule exactly; this is the first entry to exceed it, by 0.09pp.

**Suggested fix:**
Raise the branches floor to 87.0 (the nearest 0.1 that satisfies the ≤1.0pp policy) in both AGENTS.md and .github/workflows/ci.yml, or if 87.99 is itself a typo for the true CI measurement, correct the number and recompute the floor.

**Cited rule:**
> "Policy: a floor sits no more than 1.0 pp below the CI measurement for that gate; lower a floor only in the PR that causes the drop, reason in the body; raise when measured headroom exceeds 1.0 pp." (AGENTS.md, Tests section)

**Review by leehopper — COMMENTED** [review]
Code Gauntlet automated review (v3.23.1, commit 16ac5f3). 4 findings survived verification and blind-challenge: 2 medium confidence-coercion gaps in filter_findings.py (the sort-key and consensus-boost paths were not covered by this PR's own #266 fix), 1 low test-coverage gap, and 1 low floor-ratchet policy nit in AGENTS.md.
---

### ⚠️ 3 finding(s) could not be anchored inline

1 inline comment(s) were posted; the following 3 finding(s) reference lines outside this diff and are included here instead: A finding listed here may not have an anchoring problem of its own — a consolidation group whose primary could not be anchored inline is listed here in full, corroborators included.

#### `scripts/filter_findings.py:1554`

**🟡 [MEDIUM] consolidate_cross_agent's sort-key still reads raw confidence, contradicting the new #266 completeness claim**

The new #266 header comment states typed-field coercion is applied at 'every scan/concat/sort-key/threshold site that reads title/description/file/severity ... or confidence ... from a finding'. _winner_key is exactly such a sort-key site -- and was edited in this same diff to coerce desc_len via _as_text -- but its confidence read (line 1558) remains a bare .get('confidence', 0), so a null confidence still raises TypeError inside sorted()'s tuple comparison, the same crash class the PR eliminates elsewhere.

**Suggested fix:**
Wrap the confidence read with the new helper: `conf = _as_confidence(f.get("confidence"))`, matching the fix already applied to desc_len two lines below, and mirror the same change in workflows/src/filterFindings.js's winnerKey (line 1145: `const conf = pyGet(f, 'confidence', 0);`).

**Cited rule:**
> "Applied at every scan/concat/sort-key/threshold site that reads title/description/file/severity (string-typed) or confidence (numeric-typed) from a finding whose provenance is not schema-validated ... Before this, `finding.get(field, \"\")` only substitutes the default when the KEY IS ABSENT -- an explicit `null` ... passes the raw None through, which raises TypeError" (scripts/filter_findings.py, lines 113-132, #266 header comment)

#### `scripts/filter_findings.py:1554`

**💡 [LOW] consolidate_cross_agent's winner-key confidence read is left unguarded and untested for null/non-numeric confidence**

This PR hardens _winner_key's desc_len read with _as_text and adds a dedicated null-description crash-oracle fixture/test for consolidate_cross_agent, but the sibling conf = f.get("confidence", 0) on the very next line of the same tuple is left unwrapped by the new _as_confidence helper, and no test exercises a tie where two co-located, non-core-dimension findings differ only by an explicit null (or numeric-string) confidence.

**Suggested fix:**
Wrap the confidence read the same way the description read was fixed: `conf = _as_confidence(f.get("confidence"))` (and JS `const conf = asConfidence(pyGet(f, 'confidence', 0));`). Add a test with two co-located findings from different non-core dimensions (e.g. test_coverage and simplification, so is_core ties at 0) where one has `confidence: None`/`null`, asserting consolidate_cross_agent/consolidateCrossAgent does not crash and picks a deterministic winner -- mirroring the existing test_null_description_does_not_crash_winner_key / test_array_description_does_not_win_the_tiebreak_on_item_count pattern but for the confidence field.

#### `scripts/filter_findings.py:1183`

**🟡 [MEDIUM] detect_disagreement's consensus-boost/singleton-penalty logic still reads raw confidence, in the same function fixed for title/description**

detect_disagreement's Phase 2 suppression rules (lines 1109-1142, same function) were fixed in this diff to coerce title/description via _as_text. Phase 3's consensus-boost (line 1199) and singleton-penalty (line 1210) still read `finding.get("confidence", 0)` unguarded: a null confidence crashes `original_conf + _CONSENSUS_BOOST` (TypeError, None+int) or `original_conf - _SINGLETON_PENALTY`, the exact crash class #266 sets out to eliminate, left unfixed one function-section over.

**Suggested fix:**
Coerce both reads: `original_conf = _as_confidence(finding.get("confidence"))` at lines 1199 and 1210, and the matching `originalConf = asConfidence(pyGet(finding, 'confidence', 0))` in workflows/src/filterFindings.js's detectDisagreement.

**Cited rule:**
> "Applied at every scan/concat/sort-key/threshold site that reads title/description/file/severity (string-typed) or confidence (numeric-typed) from a finding whose provenance is not schema-validated ... never a stringified null, never a crash." (scripts/filter_findings.py, lines 113-132, #266 header comment)

---
Generated by code-gauntlet | Reviewed up to: 16ac5f354753fd81d5637b76c8ba6581ae14e658

**Comment by leehopper** [comment]
All review findings addressed in 25089a2:

- **Gauntlet MEDIUM ×2 + LOW**: `_winner_key`'s confidence read and `detect_disagreement`'s consensus-boost/singleton-penalty reads now route through `_as_confidence`/`asConfidence` in both twins, with crash-oracle tests (Python) and mutation-discriminating companions (JS: a numeric-string "50" confidence previously string-concatenated into a false 100 boost). The winner-key tie test covers the LOW's named gap.
- **Bugbot MEDIUM (null severity)**: the one remaining raw severity read (Phase 4 grouping, both twins) is coerced; the exact Bugbot repro (co-located pair, one null severity, through `applyFilterPipeline`/`main()`) is pinned end-to-end and was confirmed red pre-fix.
- **Gauntlet/Bugbot LOW (floor policy)**: JS branches floor corrected 86.9 → 87.0 (≤1.0pp under the #269 CI measurement 87.99) in AGENTS.md + ci.yml; local branches measure 88.10 against the new floor.
- **Bugbot MEDIUM (budget raise)**: was already addressed in 16ac5f3 (AGENTS_SET_BUDGET_BYTES 21_373 → 21_446 with the dated entry); the 86.9→87.0 edit is byte-neutral.
- **CodeQL**: redundant local imports removed.

All 8 new coercion sites mutation-verified red with byte-identical restores; suites 1153 JS / 1832 Python green; all pre-existing parity goldens byte-identical.
